content/blog/angular-extend-class/index.md

```diff
@@ -0,0 +1,842 @@
+
+```typescript
+@Component({
+  selector: 'app-root',
+  template: `
+    <p>Test</p>
+  `,
+})
+class AppComponent extends BaseComponent implements OnInit {
+  // This code doesn't work. Read on to learn why
+  constructor(@Inject(DOCUMENT) private document: Document) {
+    super(document);
+    console.log(document.body);
+  }
+}
+```
+
+But alas, this does not work!
+
+Similar to how we had to add `override` to our `AppComponent`'s lifecycle methods, we need to do the same with our constructor. Otherwise, we'll get this error:
+
+```
+TS4115: This parameter property must have an 'override' modifier because it overrides a member in base class 'BaseComponent'.
+```
+
+Let's update the code to show what that might look like:
+
+```typescript
+@Component({
+  template: ''
+})
+abstract class BaseComponent implements OnInit {
+  constructor(@Inject(DOCUMENT) private document: Document) {}
+  ngOnInit() {
+    console.log(document.title);
+  }
+}
+
+@Component({
+  selector: 'app-root',
+  template: `
+    <p>Test</p>
+  `,
+})
+class AppComponent extends BaseComponent implements OnInit {
+  constructor(@Inject(DOCUMENT) private override document: Document) {
+    super(document);
+    console.log(document.body);
+  }
+}
+```
+
+Something worth mentioning is that this code _still_ doesn't work. You'll see a compiler error with the following code:
+
+```
+TS2415: Class 'AppComponent' incorrectly extends base class 'BaseComponent'.
+   Types have separate declarations of a private property 'document'.
+```
+
+To solve this, we simply need to make our `BaseComponent`'s `constructor` properties `public` instead of `private`:
+
+```typescript
+@Component({
+  template: ''
+})
+abstract class BaseComponent implements OnInit {
+  constructor(@Inject(DOCUMENT) public document: Document) {}
+  ngOnInit() {
+    console.log(document.title);
+  }
+}
+
+@Component({
+  selector: 'app-root',
+  template: `
+    <p>Test</p>
+  `,
+})
+class AppComponent extends BaseComponent implements OnInit {
+  constructor(@Inject(DOCUMENT) public override document: Document) {
+    super(document);
+    console.log(document.body);
+  }
+}
+```
+
+> Remember to keep your `override` property in the `AppComponent` `constructor`, otherwise you'll have errors.
+
+# Why you don't want to extend Angular base classes {#dont-extend-base-classes}
+
+Now that we've learned how to extend base classes in Angular to share lifecycle methods, allow me to flip the script:
+
+**You shouldn't use a base class in Angular**.
+
+<video title="A shocked sock puppet monkey" src="./shocked-monkey-gif.mp4"></video>
+
+Why?
+
+Well, it's often sited by Angular experts that using a base class is brittle and difficult to maintain.
+
+For example, let's say that you have a base component that doesn't use dependency injection, but then suddenly need to add dependency injection. What do you do?
+
+Well, you'd have to refactor every instance that you extended that class.
```

Comment:
The `inject` function addresses this issue but there are trade-offs. It will implicitly introduce a new dependency which might:
- Result in a run-time error because of a missing provider
- Break tests because of a missing provider